src/protocols/routing/mod.rs

```diff
@@ -0,0 +1,270 @@
+
+    if service.is_none() {
+        return Ok(vec![]);
+    }
+
+    let mut service = service.unwrap();
+
+    let mut services = vec![service.clone()];
+    let mut service_endpoint = &unwrap_did_comm_service(&service)?.service_endpoint;
+
+    while is_did(service_endpoint) {
+        // Now alternative endpoints recursion is not supported
+        // (because it should not be used according to the specification)
+        if services.len() > 1 {
+            return Err(err_msg(
+                ErrorKind::InvalidState,
+                "DID doc defines alternative endpoints recursively",
+            ));
+        }
+
+        service = find_did_comm_service(service_endpoint, None, did_resolver)
+            .await?
+            .ok_or_else(|| {
+                err_msg(
+                    ErrorKind::InvalidState,
+                    "Referenced mediator does not provide any DIDCommMessaging services",
+                )
+            })?;
+        services.insert(0, service.clone());
+        service_endpoint = &unwrap_did_comm_service(&service)?.service_endpoint;
+    }
+
+    Ok(services)
+}
+
+fn generate_message_id() -> String {
+    Uuid::new_v4().to_string()
+}
+
+fn build_forward_message(
+    forwarded_msg: &str,
+    next: &str,
+    headers: Option<&Vec<(String, Value)>>,
+) -> Result<String> {
+    let body = json!({ "next": next });
+
+    let attachment = Attachment::json(serde_json::from_str(forwarded_msg)?).finalize();
+
+    let mut msg_builder = Message::build(generate_message_id(), FORWARD_MSG_TYPE.to_owned(), body);
+
+    if let Some(headers) = headers {
+        for (name, value) in headers {
+            msg_builder = msg_builder.header(name.to_owned(), value.to_owned());
+        }
+    }
+
+    msg_builder = msg_builder.attachment(attachment);
+
+    let msg = msg_builder.finalize();
+
+    serde_json::to_string(&msg).kind(ErrorKind::InvalidState, "Unable serialize forward message")
+}
+
+pub fn try_parse_forward(msg: &Message) -> Option<ParsedForward> {
+    if msg.type_ != FORWARD_MSG_TYPE {
+        return None;
+    }
+
+    let next = match msg.body {
+        Value::Object(ref body) => match body.get("next") {
+            Some(&Value::String(ref next)) => Some(next),
+            _ => None,
+        },
+        _ => None,
+    };
+
+    if next.is_none() {
+        return None;
+    }
+
+    let next = next.unwrap();
+
+    let json_attachment_data = match msg.attachments {
+        Some(ref attachments) => match &attachments[..] {
+            [attachment, ..] => match &attachment.data {
+                AttachmentData::Json(forwarded_msg_data) => Some(forwarded_msg_data),
+                _ => None,
+            },
+            _ => None,
+        },
+        None => None,
+    };
+
+    if json_attachment_data.is_none() {
+        return None;
+    }
+
+    let forwarded_msg = &json_attachment_data.unwrap().json;
+
+    Some(ParsedForward {
+        msg: msg.clone(),
```

Comment:
Can we avoid cloning here? Can we reference an original message in ParsedForward, or borrow the original message (assuming it's not needed anymore and ParsedForward will be used instead)?